Changes in module 'nf-core/star/genomegenerate'
--- modules/nf-core/star/genomegenerate/main.nf
+++ modules/nf-core/star/genomegenerate/main.nf
@@ -2,18 +2,19 @@
     tag "$fasta"
     label 'process_high'
 
-    conda "bioconda::star=2.7.10a bioconda::samtools=1.16.1 conda-forge::gawk=5.1.0"
+    // Note: 2.7X indices incompatible with AWS iGenomes.
+    conda "bioconda::star=2.7.9a bioconda::samtools=1.15.1 conda-forge::gawk=5.1.0"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
-        'https://depot.galaxyproject.org/singularity/mulled-v2-1fa26d1ce03c295fe2fdcf85831a92fbcbd7e8c2:1df389393721fc66f3fd8778ad938ac711951107-0' :
-        'quay.io/biocontainers/mulled-v2-1fa26d1ce03c295fe2fdcf85831a92fbcbd7e8c2:1df389393721fc66f3fd8778ad938ac711951107-0' }"
+        'https://depot.galaxyproject.org/singularity/mulled-v2-1fa26d1ce03c295fe2fdcf85831a92fbcbd7e8c2:1c4c32d87798d425c970ececfbadd155e7560277-0' :
+        'quay.io/biocontainers/mulled-v2-1fa26d1ce03c295fe2fdcf85831a92fbcbd7e8c2:1c4c32d87798d425c970ececfbadd155e7560277-0' }"
 
     input:
     path fasta
     path gtf
 
     output:
-    path "star"        , emit: index
-    path "versions.yml", emit: versions
+    path "star"         , emit: index
+    path "versions.yml" , emit: versions
 
     when:
     task.ext.when == null || task.ext.when
@@ -21,7 +22,7 @@
     script:
     def args = task.ext.args ?: ''
     def args_list = args.tokenize()
-    def memory = task.memory ? "--limitGenomeGenerateRAM ${task.memory.toBytes() - 100000000}" : ''
+    def memory   = task.memory ? "--limitGenomeGenerateRAM ${task.memory.toBytes() - 100000000}" : ''
     if (args_list.contains('--genomeSAindexNbases')) {
         """
         mkdir star
@@ -33,7 +34,6 @@
             --runThreadN $task.cpus \\
             $memory \\
             $args
-
         cat <<-END_VERSIONS > versions.yml
         "${task.process}":
             star: \$(STAR --version | sed -e "s/STAR_//g")
@@ -45,7 +45,6 @@
         """
         samtools faidx $fasta
         NUM_BASES=`gawk '{sum = sum + \$2}END{if ((log(sum)/log(2))/2 - 1 > 14) {printf "%.0f", 14} else {printf "%.0f", (log(sum)/log(2))/2 - 1}}' ${fasta}.fai`
-
         mkdir star
         STAR \\
             --runMode genomeGenerate \\
@@ -56,7 +55,6 @@
             --genomeSAindexNbases \$NUM_BASES \\
             $memory \\
             $args
-
         cat <<-END_VERSIONS > versions.yml
         "${task.process}":
             star: \$(STAR --version | sed -e "s/STAR_//g")
@@ -65,32 +63,4 @@
         END_VERSIONS
         """
     }
-
-    stub:
-    """
-    mkdir star
-    touch star/Genome
-    touch star/Log.out
-    touch star/SA
-    touch star/SAindex
-    touch star/chrLength.txt
-    touch star/chrName.txt
-    touch star/chrNameLength.txt
-    touch star/chrStart.txt
-    touch star/exonGeTrInfo.tab
-    touch star/exonInfo.tab
-    touch star/geneInfo.tab
-    touch star/genomeParameters.txt
-    touch star/sjdbInfo.txt
-    touch star/sjdbList.fromGTF.out.tab
-    touch star/sjdbList.out.tab
-    touch star/transcriptInfo.tab
-
-    cat <<-END_VERSIONS > versions.yml
-    "${task.process}":
-        star: \$(STAR --version | sed -e "s/STAR_//g")
-        samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//')
-        gawk: \$(echo \$(gawk --version 2>&1) | sed 's/^.*GNU Awk //; s/, .*\$//')
-    END_VERSIONS
-    """
-}
+}
************************************************************
