Changes in module 'nf-core/samtools/sort'
--- modules/nf-core/samtools/sort/main.nf
+++ modules/nf-core/samtools/sort/main.nf
@@ -20,7 +20,7 @@
 
     script:
     def args = task.ext.args ?: ''
-    def prefix = task.ext.prefix ?: "${meta.id}"
+    def prefix = task.ext.prefix ?: "task.ext.prefix"
     if ("$bam" == "${prefix}.bam") error "Input and output names are the same, use \"task.ext.prefix\" to disambiguate!"
     """
     samtools sort $args -@ $task.cpus -o ${prefix}.bam -T $prefix $bam

************************************************************
